### PR TITLE
feat: Check for updates and display a warning when an update is available 

### LINKED ISSUE #299: Look for updates and alert the user when available
**Is your feature request related to a problem? Please describe.**
Users are not informed when a new version is released, unless they watch the repository on GitHub.

**Describe the solution you'd like**
An optional feature to look for the version number of the latest version and print a warning when a newest version is available when ggshield is executed.

**Describe alternatives you've considered**
A new command to look for update, it would be less intrusive but only alert users who go out of their way to use it.

### PR DESCRIPTION
**Pull request by Walz**
When the command is called, look for updates by requesting the latest release on GitHub. Save the latest version in a local file to only query GitHub once per day.

image

closes #299

### CODE REVIEW

File: tests/unit/core/test_utils.py [hunk 1]
```diff
@@ -330,3 +337,67 @@ def test_load_dot_env_loads_git_root_env(
...
+    ),
+)
+def test_check_for_update(
+    request_get_mock: Mock,
+    current_version: str,
+    remote_version: str,
+    print_warning: bool,
+    capsys: CaptureFixture[str],
+    monkeypatch: MonkeyPatch,
+):
```
Comment (agateau-gg) [inline tests/unit/core/test_utils.py:362]: You can use the `fs` fixture from [pyfakefs](https://pypi.org/project/pyfakefs/) in your test to ensure the test never touches the real file system. This avoids side-effects and removes the need for the optional `cache` argument in `check_for_update()`.

File: ggshield/core/utils.py [hunk 2]
```diff
@@ -388,3 +393,53 @@ class ScanContext:
...
+    if (
+        latest_version is None
+        or check_at < 0
+        or (time.time() - check_at > 24 * 60 * 60)
+    ):
+        logger.debug("Checking the latest released version of ggshield...")
+        resp = requests.get(
+            "https://api.github.com/repos/GitGuardian/GGShield/releases/latest"
+        )
+        if resp.status_code == 200:
```
[the remark this answers is not in the document]
Comment (Walz) [reply ggshield/core/utils.py]: Even if the request to GitHub API fails, we may still return a warning if a new version has been cached.

**Review by agateau-gg — APPROVED** [review]
Looks good but... CI is not happy 😢 

I suspect the failures come from the fact that we execute the unit-test test-suite with `--disable-socket`, and since a lot of tests go through `cli()` we end up checking for updates a lot :).

This is, I think, a rare case where it makes sense for the main code to alter its behavior depending on whether it's being run by pytest. I think you can fix this by modifying `main.cli()` to not check for updates if the `PYTEST_CURRENT_TEST` environment variable is set. This looks a nice way to detect if the code is running under pytest without importing pytest in the code itself.

File: ggshield/cmd/main.py [hunk 3]
```diff
@@ -140,6 +148,11 @@ def cli(
     logger.debug("args=%s", sys.argv)
     logger.debug("py-gitguardian=%s", pygitguardian.__version__)
 
+    if check_for_updates and "PYTEST_CURRENT_TEST" not in os.environ:
```
Comment (agateau-gg) [inline ggshield/cmd/main.py:151]: Can you add a comment explaining why we added the `PYTEST_CURRENT_TEST` check? Should help future-us being less puzzled about this code.

File: ggshield/core/check_updates.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                {"latest_version": latest_version, "check_at": check_at}, CACHE_FILE
+            )
+
+    if latest_version is not None:
+        current_version_split = tuple(map(int, __version__.split(".")))
+        latest_version_split = tuple(map(int, latest_version.split(".")))
+        if current_version_split < latest_version_split:
+            return (
+                f"A new version of ggshield (v{latest_version}) has been released "
+                f"(https://github.com/GitGuardian/ggshield)."
```
Comment (agateau-gg) [inline ggshield/core/check_updates.py:64]: That's not exactly what I had in mind with "separation of concerns": the idea is this function does not need to concern itself with how the information is going to be presented to the user: it should return either None (no update available), or the version number of the available version. Then it's up to the caller to present the information, which in our case means including the version number in a text message.

**Review by agateau-gg — CHANGES_REQUESTED** [review]
Minor remarks. Looks good otherwise.

**Review by agateau-gg — APPROVED** [review]
Looks good, thanks!

**Pull request merged**